### PR TITLE
Add stat collection to IPTB

### PR DESCRIPTION
**Pull request by davinci26**
Hey y'all,

Thanks for the project it helped me a lot!

As discussed in IPFS issue board [IPFS Performance #5226](https://github.com/ipfs/go-ipfs/issues/5226), I made some changes to the IPTB framework to generate measure the performance of IPFS and generate performance graphs. In detail I added the following functions in the framework:
- iptb make-topology: This creates a connection graph between the nodes (e.g. star topology, barbell topology). In the topology files empty lines and lines starting with # are disregarded. For non empty line the syntax is `origin:connection 1, connection 2 ...` where origin and connections are specified with their node ID.
- iptb dist -hash: The simulation here distributes a single file from node 0 to every other node in the network. Then it calculates the average time required to download the file, the standard deviation of the time, the maximum time, the minimum time, the duplicate blocks. The results are saved in a generated file called results.json. 

I also added a Python3 script to plot the results that adds an additional optional dependency in the project, Matplotlib. 

Finally I created a readme file, simulation.md, that explains the logic of the simulation. I also added there the response of @whyrusleeping  to the issue [Simulate bad network #50](https://github.com/ipfs/iptb/issues/50) so people would know that there is support for bad network.

I would appreciate your feedback and any improvements suggestions :)

### CODE REVIEW

**Comment by whyrusleeping** [comment]
Hey @davinci26 Thanks a bunch for this! I hope you don't mind us pulling this apart a bit. We're working on refactoring iptb to be less ipfs centric (giving it the ability to really test *any* libp2p based protocol). And to complete that work, it may require some cutting and pasting :)

I'll leave it to @travisperson and @frrist to take it from here and see how integrating this work could happen.

**Comment by davinci26** [comment]
No problem, feel free to make any changes you want!

I spent quite some time playing with this repo so if you want any help in the refactoring or in the integration, give me a shout :)

**Comment by travisperson** [comment]
This is really cool and highlights what iptb can do nicely. 🏆 

You can checkout the current work we are doing in PR #61, and there is an issue which has some discussion in #62 if you want to see what changes we are making. My comments below will be with regard to this work and how we can incorporate the changes you have here.

I think we can probably incorporate the idea behind these changes and improve iptb.

It looks like `make-topology` is a nice configuration based way to run a bunch of `iptb connect` calls which I can see having value when running iptb from an interactive shell. It would be interesting to explore ways to make some of it more generic, though I don't think that would be at all required for a first step.

I could see us adding a new flag to [`connect`](https://github.com/ipfs/iptb/pull/61/files#diff-1b79de1ed0f248cd2a9f0df5545cad39), which would specify a topology file. If the flag is specified, then the arguments to connect would no longer be required, and the file would be used to run the connect calls.

`iptb connect --topology /tmp/barbell.txt`

I really like the idea of baking stats into iptb, as it seems like a perfect use case. I think we could make it more generic as well, enabling it to be used to record a lot of different commands, for example, targeting the dht specifically.

We could make the `run` command be able to record basic statistics about command execution by passing a `--stats` flag, which specified a file to write stats out too, no arguments would dump the stats to stdout.

The `dist` command as written now, could be ran with something like `iptb run --stats results.json -- ipfs cat <hash>`.

_Note: `iptb run -- <cmd>`, runs `<cmd>` on all nodes, it replaces `for-each`._

(If we wanted to run it against a subset of nodes, we could do `iptb run --stats results.json [1,3,5-9] -- ipfs cat $HASH`).

The only stat that does not carry over outside of ipfs very well is the `duplicate block` stat. That particular stat though I think is a perfect candidate for the new `metric` part of the api. We could make the stats take a list of metrics to collect along side the standard time execution stuff.

`iptb run [--stats <outfile>] [--collect <metric> ...] -- <cmd>`

**Comment by davinci26** [comment]
Thanks for the review! I totally agree with your suggestions.

From my experience, the problem I had with the `For each` cli command was that it was synchronous. This prohibited me from simulating multiple nodes trying to do the same thing.

Maybe this could also be added as an optional flag in the run command:
`iptb run --async --stats results.json -- ipfs cat <hash> `

I can add the changes as you described them after you merge PR #61 to master, or you can add them directly to PR #61 :).

**Comment by davinci26** [comment]
Hey y'all,
I saw that PR #61 was merged into master.

Can I start working on the `topology` feature and the `statistics metrics` based on the suggestions and comments of @travisperson?

**Comment by davinci26** [comment]
Hey again,

I made the majority of the changes so feel free to review the code. I also added unit tests on the Stats just to be sure there is no stupid miscalculation.

The stats produced are the following:

```
type Stats struct {
	Mean      float64
	Std       float64
	Max       float64
	Quantile3 float64
	Median    float64
	Quantile1 float64
	Min       float64
}

```
Stats Implementation details:

1. In the function `mapWithOutput` I calculate the time elapsed for the command and use it to add it in the `Results struct`
2. I added in the `Build Report` a flag to calculate the stats if the user requests it. Currently, the stats are just printed on the std out. If you are happy with the design I can modify it to also be printed in a file if specified by the user. I print the Stats in `JSON` format for easier parsing by other programs.
3. Also in the `Build Report` I pass an additional string which is the command executed. For me it feels more correct in the case that the output will be automatically parsed by another program to produce statistics (calculate bottlenecks on pipelines). 
4. Right now the majority of commands produce the stats. This could easily be disabled if you dont think its useful.
5.  I assume it would be useful to add stats besides time on the various plugins. However, I am not 100% sure what kind of metrics the plugins you plan to add produce. If you have any reading material on that then I can take a look into it.

Topology Implementation details:

1. Added one function that parses a topology files with the same format as above and then transforms it into the `range format` used internally.

Looking forward to your comments 😃

File: commands/utils.go [hunk 1]
```diff
@@ -225,91 +226,73 @@ func buildReport(results []Result, command string, statsFlag bool) error {
...
-	c1 := 0
-	c2 := 0
-	if length%2 == 0 {
-		c1 = length/2 + 1
-		c2 = length/2 - 1
-	} else {
-		c1 = (length-1)/2 + 1
-		c2 = c1 - 1
+// Print individual results to JSON format
+func printIndivudualResults(nodeTime []float64, nodeID []int, metric string) {
```
Comment (dgrisham) [inline commands/utils.go]: Think that's a typo and should be `printIndividualResults` ;)

File: commands/utils.go [hunk 2]
```diff
@@ -195,13 +203,96 @@ func buildReport(results []Result) error {
...
+}
+
+/*
+Substract_arrays
+Substract y from x returns [xi-yi] for i in len(x)
+If the arrays do not have the same length or
+the strings are not castable to float
+the function returns an error
+*/
+func substractArrays(lhs []string, rhs []string) ([]float64, error) {
```
Comment (dgrisham) [inline commands/utils.go]: One more typo, `subtractArrays`*

File: commands/stats.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+//BuildStats generates a stats struct from an input float64 slice
+//It returns an error if the slice is empty
+func BuildStats(results []float64) (Stats, error) {
+	if len(results) == 0 {
+		return Stats{}, errors.New("Results are empty")
+	}
+
+	// Calculate the min, max mean, std with a single pass of the array
+	sum := 0.
+	sumSquarred := 0.
```
Comment (dgrisham) [inline commands/stats.go]: `sumSquared` (just one r)

File: commands/stats.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+	Std       float64
+	Max       float64
+	Quantile3 float64
+	Median    float64
+	Quantile1 float64
+	Min       float64
+}
+
+//BuildStats generates a stats struct from an input float64 slice
+//It returns an error if the slice is empty
```
Comment (dgrisham) [inline commands/stats.go]: These two lines are missing a space after the `//` to be consistent with the others (sorry, details stand out to me haha)

File: README.md [hunk 5]
```diff
@@ -26,7 +26,7 @@ $ iptb shell 4
 $ ipfs cat QmNqugRcYjwh9pEQUK7MLuxvLjxDNZL1DH8PJJgWtQXxuF
 hey!
 ```
-
+Available plugins now: Local IPFS node (plugin_name: localipfs), Docker IPFS node (plugin_name: dockeripfs)
```
Comment (dgrisham) [inline README.md]: Would also be nice to be able to list available plugins via the CLI. Doesn't have to be in this PR though, just mentioning.

File: commands/connect.go [hunk 6]
```diff
@@ -23,13 +27,17 @@ The connect command allows for connecting sets of nodes together.
 Every node listed in the first set, will try to connect to every node
 listed in the second set.
 
-There are three variants of the command. It can accept no arugments,
-a single argument, or two arguments. The no argument and single argument
-expands out to the two argument usage.
+There are four variants of the command. It can accept no arugments,
```
Comment (dgrisham) [inline commands/connect.go]: arguments*

File: commands/connect.go [hunk 7]
```diff
@@ -23,13 +27,17 @@ The connect command allows for connecting sets of nodes together.
...
+a single argument, two arguments or a file. The no argument and single argument
+expands out to the two argument usage. The file argument, parses the file and exapnds it
```
Comment (dgrisham) [inline commands/connect.go]: expands* (the second one)

File: commands/utils.go (hunk 2 above)
Comment (davinci26) [reply commands/utils.go]: I will fix all of them :) Thnx a lot for the review, I totally agree details matter a lot !

Comment (dgrisham) [reply commands/utils.go]: No worries! And one of the comments I made got marked as outdated for some reason, but `printIndividualResults` is typo'd as `printIndivudualResults`

File: main.go [hunk 8]
```diff
@@ -14,6 +14,7 @@ import (
 )
 
 func loadPlugins(dir string) error {
+
```
[the remark this answers is not in the document]
Comment (davinci26) [reply main.go]: The code formatting tool decided to add an empty line there. I have manually checked the files and they are the same.

File: commands/run.go [hunk 9]
```diff
@@ -54,11 +65,51 @@ var RunCmd = cli.Command{
...
+		}
+		value, err := metricNode.Metric(metric)
+		if err != nil {
+			return nil, err
+		}
+
+		nodeMetricValue[i] = value
+	}
+	return nodeMetricValue, nil
+}
```
Comment (davinci26) [inline commands/run.go]: Here I opted to collect the metric before and after the command is executed in `run.go`. I think with this approach the code is cleaner and makes more sense.
I could add it in the `mapWithOutput` function but since this function is using locks the more operations I do inside the locks the more the overhead.

I would love to hear your thoughts on which design you think is more useful for the metrics you plan to implement.

File: commands/utils.go [hunk 10]
```diff
@@ -195,13 +203,97 @@ func buildReport(results []Result) error {
...
+
+/*
+Substract_arrays
+Substract y from x returns [xi-yi] for i in len(x)
+If the arrays do not have the same length or
+the strings are not castable to float
+the function returns an error
+TODO: This could be more generic
+*/
+func subtractArrays(lhs []string, rhs []string) ([]float64, error) {
```
Comment (davinci26) [inline commands/utils.go]: As it is implemented now the metrics are collected as follows:
Let metric <m1> be collected, then we collect m1<sub>before</sub> before the execution of the run command and m1<sub>after</sub> after. Then we calculate the statistics based on Δm1 = (m1<sub>after</sub> - m1<sub>before</sub>).

**Comment by travisperson** [comment]
I was giving this some thought over the weekend and I think we should try and only add what is minimally required to expose the functionality required to do these stat calculations. While iptb could actually do all the calculations, I don't know if there is a strong reason why it needs to do them. It might be better to just report the raw stats and allow the user to handle how to calculate values from them. So with this we would just be adding an elapsed time to the results (as you already have).

Collecting metrics before and after isn't a lot different than running `iptb metric <node> <metric>` manually before or after (the command probably should support ranges), so I don't know if we should add anything like that at the moment.

Per our discussion on IRC, you also pointed out that for metrics it would be better to actually sample over the course of the commands execution. I think this would probably be the correct thing to do as well for any realtime stats (cpu / memory / etc). This does get a little bit complicated though and I think we really need to start a discussion in another issue and work through exactly the requirements needed to find a good approach if we want to go down this route.

For this PR though I think we can keep it pretty simple, and just always report the elapse time for all commands ran that use the standard `buildReport`. We will no longer need the --stats flag right now.

```
iptb run -- ipfs id --format="<id>"
node[0] exit 0; elapsed 10ns

QmXschyVzLmS4JqPN1kuhCXTjau2oQkVuzjvTbQFTGm3w3
node[1] exit 0; elapsed 10ns

Qme3h8WwfpBiPHPfdEs9GuegijVhaBX9xYPXTTDAS6uciR
node[2] exit 0; elapsed 10ns

QmbKpZTyRodMmjkBPdyngkha1ic6bu7YVXPzisRdFNi5Yq
```

To summarize. In this PR we should have the connect topology, and in addition add a elapsed time to all results / buildReport.

I think we can easily move forward with that, and then start to discuss further how we can approach the other topics.

**Comment by travisperson** [comment]
On an additional note I'm actually going to push back on the topology feature as well right now.

Sorry for this being a bit messy. I should have requested an issue from the start so that we could have a discussion outside of the code.

My concerns at the moment with regard to the topology is that it can be accomplished with a shell script fairly easily, and I don't see the exact benefits of the topology file at the moment due to the restrictiveness of the format, as it works with exactly N nodes.

Let take a step back from this and open two issues to see how we can address the requests

1) Connect topology
2) Collecting statistics

I'll open these issues and bring some context in, we can then discuss them from there.

**Comment by davinci26** [comment]
The last commit reflects the discussion in #77 #76 . The user can run 
`ipfs run 0 --encoding JSON <args>` or `ipfs run 0 --encoding text <args>` and get the outpout in the said encoding.
The JSON encoding transforms the output (including the plugin stderr and stdout) into a JSON. This way the user can specify he/she prefers a human-readable or a machine-readable output.

File: commands/run.go [hunk 11]
```diff
@@ -21,6 +22,11 @@ var RunCmd = cli.Command{
 			Name:   "terminator",
 			Hidden: true,
 		},
+		cli.StringFlag{
+			Name:  "encoding",
+			Usage: "Specify the output format, current options JSON and text",
+			Value: "text",
+		},
```
Comment (travisperson) [inline commands/run.go]: We should make this a global flag under the app in main.go so that it works for other commands. The format error checking below can take place in the `Before` block as well.

File: commands/utils.go [hunk 12]
```diff
@@ -121,9 +124,19 @@ func expandDashRange(s string) ([]int, error) {
...
+	Error   error
+	Elapsed float64
+}
+
+type output struct {
+	Node         int
+	ExitCode     int
+	Errors       error
+	PluginStdOut string
+	PluginStdErr string
```
Comment (travisperson) [inline commands/utils.go]: `Errors` => `Error` as it's a single error.

nit: To keep the casing similar to other usages through the code and go in general I think we should use `Stdout` and `Stderr` here.

File: commands/utils.go [hunk 13]
```diff
@@ -184,21 +199,36 @@ func buildReport(results []Result) error {
...
+				io.Copy(os.Stdout, rs.Output.Stderr())
+			} else {
+				pluginOut := ""
+				pluginErr := ""
+				if b, err := ioutil.ReadAll(rs.Output.Stdout()); err == nil {
+					pluginOut = string(b)
+				}
+				if b, err := ioutil.ReadAll(rs.Output.Stderr()); err == nil {
+					pluginErr = string(b)
+				}
```
Comment (travisperson) [inline commands/utils.go]: We need to handle the other side of this error, when the error is not nil.

File: commands/utils.go [hunk 14]
```diff
@@ -184,21 +199,36 @@ func buildReport(results []Result) error {
...
+				rsJSON, _ := json.Marshal(output{
+					Node:         rs.Node,
+					ExitCode:     rs.Output.ExitCode(),
+					Errors:       rs.Output.Error(),
+					PluginStdOut: pluginOut,
+					PluginStdErr: pluginErr,
+					Elapsed:      rs.Elapsed,
+				})
```
Comment (travisperson) [inline commands/utils.go]: Need to also handle this error.

File: commands/utils.go [hunk 15]
```diff
@@ -184,21 +199,41 @@ func buildReport(results []Result) error {
...
+				}
+				pluginOut := string(pluginOutB)
+				// Transform plugin stderr to string
+				pluginErrB, err := ioutil.ReadAll(rs.Output.Stderr())
+				if err != nil {
+					errs = append(errs, err)
+				}
+				pluginErr := string(pluginErrB)
+				// JSONify the plugin output
+				rsJSON, _ := json.Marshal(output{
```
Comment (travisperson) [inline commands/utils.go]: missing this error

File: commands/utils.go [hunk 16]
```diff
@@ -184,21 +199,41 @@ func buildReport(results []Result) error {
...
+				})
+				fmt.Printf("%s\n", rsJSON)
 			}
-
-			fmt.Println()
-
-			io.Copy(os.Stdout, rs.Output.Stdout())
-			io.Copy(os.Stdout, rs.Output.Stderr())
-
 			fmt.Println()
```
Comment (travisperson) [inline commands/utils.go]: We should move this up inside of the text encoding section, and then also make sure we have a newline outside of the loop (it can be printed for all cases that are not text).

This will make it so that all output for a single command will be ld-json, and between each execution will be a newline. This will help break up the output if appending to the same file.

The output should look like if it where being appended to the same file

```
{"Node": 0, ...}
{"Node": 1, ...}
{"Node": 2, ...}

{"Node": 0, ...}
{"Node": 1, ...}
{"Node": 2, ...}
```

**Comment by davinci26** [comment]
I handled the error cases mentioned and also the formatting is as specified when I append to a file.

**Comment by davinci26** [comment]
Hey @frrist,

I was thinking about this as well, that this PR is getting too big and for no reason, as there was no design discussion prior to the beginning of the work.  It was also an omission from my side for not opening an issue prior to the PR. Feel free to decline the PR.

I have written my ideas in #77 under general thoughts and I am more than happy to help in designing meaningful event logs.

**Comment by whyrusleeping** [comment]
I'm really excited that iptb is getting so much attention. I love a lot of the direction here, let's get things designed nicely and march onwards :) Thanks @davinci26 for pushing!

One thing I always try to keep in mind in the design of tools is the unix philosophy, primarily "do one thing well". What that means here specifically, I'm not sure, but I'm thinking that we may end up adding a companion tool to iptb for metrics collection, and just ensuring that iptb doesnt prevent us from getting the information we need. This would allow us to more easily tune the stats we collect for each usecase. (similar design philosophy went into the plugins architecture)
